From 01fe90a34c6496088cedd16032e3aab7be4de0ff Mon Sep 17 00:00:00 2001
From: Amelie Delaunay <amelie.delaunay@foss.st.com>
Date: Tue, 8 Mar 2022 08:56:01 +0100
Subject: [PATCH 11/49] dmaengine: stm32-dma: fix chaining end of transfer
 management

If chaining is used and the direction is MEM_TO_MEM or DEV_TO_MEM, the
transfer is completed when MDMA has finished to push data in memory,
reported by stm32_mdma_chan_complete() callback. So it is up to the
callback to modify chan->busy and chan->status.

Change-Id: Iecd45a1c5b52d8259d1074fb52bed18ea1d75f76
Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/linux.git 8d6ac6b8dfa13b43c81257ac490d2855a6a1677a]
Fixes: 73c0a39c5d48 ("dmaengine: stm32-dma: Add DMA/MDMA chaining support")
Signed-off-by: Amelie Delaunay <amelie.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/242858
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Fabien DESSENNE <fabien.dessenne@foss.st.com>
---
 drivers/dma/stm32-dma.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/dma/stm32-dma.c b/drivers/dma/stm32-dma.c
index 1d2af6e31d0ab..f3c824917dc39 100644
--- a/drivers/dma/stm32-dma.c
+++ b/drivers/dma/stm32-dma.c
@@ -1109,10 +1109,10 @@ static void stm32_dma_handle_chan_done(struct stm32_dma_chan *chan, u32 scr)
 		else if (scr & STM32_DMA_SCR_DBM)
 			stm32_dma_configure_next_sg(chan);
 	} else {
+		if (chan->use_mdma && chan->mchan.dir != DMA_MEM_TO_DEV)
+			return; /* wait for callback */
 		chan->busy = false;
 		chan->status = DMA_COMPLETE;
-		if (chan->use_mdma && chan->mchan.dir != DMA_MEM_TO_DEV)
-			return;
 		if (chan->next_sg == chan->desc->num_sgs) {
 			vchan_cookie_complete(&chan->desc->vdesc);
 			chan->desc = NULL;
-- 
2.39.2

